plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoTypes.java

```diff
@@ -0,0 +1,173 @@
+     *
+     * @throws IllegalStateException if this type is not {@link Type#isOrderable() orderable}
+     */
+    public static Optional<Object> getPreviousValue(Type type, Object value)
+    {
+        if (!type.isOrderable()) {
+            throw new IllegalStateException("Type is not orderable: " + type);
+        }
+        requireNonNull(value, "value is null");
+
+        if (type == TINYINT || type == SMALLINT || type == INTEGER || type == BIGINT) {
+            Range typeRange = type.getRange().orElseThrow();
+            return getAdjacentValue((long) typeRange.getMin(), (long) typeRange.getMax(), (long) value, Direction.PREV);
+        }
+
+        if (type == DATE) {
+            // TODO update the code here when type implements getRange
+            verify(type.getRange().isEmpty(), "Type %s unexpectedly returned a range", type);
+            return getAdjacentValue(Integer.MIN_VALUE, Integer.MAX_VALUE, (long) value, Direction.PREV);
+        }
+
+        if (type instanceof TimestampType) {
+            // Iceberg supports only timestamp(6)
+            checkArgument(((TimestampType) type).getPrecision() == 6, "Unexpected type: %s", type);
+            // TODO update the code here when type implements getRange
+            verify(type.getRange().isEmpty(), "Type %s unexpectedly returned a range", type);
+            return getAdjacentValue(Long.MIN_VALUE, Long.MAX_VALUE, (long) value, Direction.PREV);
+        }
+
+        if (type instanceof TimestampWithTimeZoneType) {
+            // Iceberg supports only timestamp(6)
+            checkArgument(((TimestampWithTimeZoneType) type).getPrecision() == 6, "Unexpected type: %s", type);
+            verify(type.getRange().isEmpty(), "Type %s unexpectedly returned a range", type);
+            LongTimestampWithTimeZone timestampTzValue = (LongTimestampWithTimeZone) value;
+            long epochMillis = timestampTzValue.getEpochMillis();
+            int picosOfMilli = timestampTzValue.getPicosOfMilli();
+            picosOfMilli -= PICOSECONDS_PER_MICROSECOND;
+            if (picosOfMilli < 0) {
+                if (epochMillis == Long.MIN_VALUE) {
+                    return Optional.empty();
+                }
+                epochMillis--;
+                picosOfMilli += PICOSECONDS_PER_MILLISECOND;
+            }
+            // The zone doesn't matter for timestamp with time zone comparisons. Use UTC to avoid confusion.
+            return Optional.of(LongTimestampWithTimeZone.fromEpochMillisAndFraction(epochMillis, picosOfMilli, UTC_KEY));
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Returns the minimum value that compares greater than {@code value}.
+     * <p>
+     * The type of the value must match {@link Type#getJavaType}.
+     *
+     * @throws IllegalStateException if this type is not {@link Type#isOrderable() orderable}
+     */
+    public static Optional<Object> getNextValue(Type type, Object value)
```

Comment:
if you follow the java Math library naming conventions, this would be called: nextAfter or nextUp.

Reply:
( same answer as in https://github.com/trinodb/trino/pull/12795#discussion_r901523225 )